File: server/services/projects/projectService.js
// services/projects/projectService.js - 项目服务
import { Project } from '../../models/index.js';

const ProjectService = {
  /**
   * 获取所有项目
   * @param {object} db - 数据库连接对象
   * @returns {Promise} - 返回查询结果的Promise
   */
  getAllProjects: async (db) => {
    try {
      return await Project.findAll(db);
    } catch (error) {
      console.error('获取项目列表失败:', error);
      throw error;
    }
  },

  /**
   * 根据ID获取项目
   * @param {object} db - 数据库连接对象
   * @param {number} id - 项目ID
   * @returns {Promise} - 返回查询结果的Promise
   */
  getProjectById: async (db, id) => {
    try {
      return await Project.findById(db, id);
    } catch (error) {
      console.error(`获取项目 ${id} 失败:`, error);
      throw error;
    }
  },

  /**
   * 创建项目
   * @param {object} db - 数据库连接对象
   * @param {object} projectData - 项目数据
   * @returns {Promise} - 返回创建结果的Promise
   */
  createProject: async (db, projectData) => {
    try {
      // 验证必要字段
      if (!projectData.name || !projectData.start_date) {
        throw new Error('项目名称和开始日期是必填项');
      }

      // 如果没有提供状态，默认为'进行中'
      if (!projectData.status) {
        projectData.status = '进行中';
      }

      const projectId = await Project.create(db, projectData);
      return await Project.findById(db, projectId);
    } catch (error) {
      console.error('创建项目失败:', error);
      throw error;
    }
  },

  /**
   * 更新项目
   * @param {object} db - 数据库连接对象
   * @param {number} id - 项目ID
   * @param {object} projectData - 项目数据
   * @returns {Promise} - 返回更新结果的Promise
   */
  updateProject: async (db, id, projectData) => {
    try {
      // 验证必要字段
      if (!projectData.name || !projectData.start_date) {
        throw new Error('项目名称和开始日期是必填项');
      }

      const changes = await Project.update(db, id, projectData);
      if (changes === 0) {
        throw new Error(`项目 ${id} 不存在`);
      }
      
      return await Project.findById(db, id);
    } catch (error) {
      console.error(`更新项目 ${id} 失败:`, error);
      throw error;
    }
  },

  /**
   * 删除项目
   * @param {object} db - 数据库连接对象
   * @param {number} id - 项目ID
   * @returns {Promise} - 返回删除结果的Promise
   */
  deleteProject: async (db, id) => {
    try {
      const changes = await Project.delete(db, id);
      if (changes === 0) {
        throw new Error(`项目 ${id} 不存在`);
      }
      return { message: '项目删除成功' };
    } catch (error) {
      console.error(`删除项目 ${id} 失败:`, error);
      throw error;
    }
  }
};

export default ProjectService;

File: src-tauri/src/services/project_service.rs
use crate::models::project::{Project, CreateProjectDto, UpdateProjectDto};
use sqlx::SqlitePool;

pub async fn get_all_projects(pool: &SqlitePool) -> Result<Vec<Project>, String> {
    let projects = sqlx::query_as::<_, Project>(
        "SELECT * FROM projects ORDER BY created_at DESC"
    )
    .fetch_all(pool)
    .await
    .map_err(|e| format!("获取项目列表失败: {}", e))?;
    
    Ok(projects)
}

pub async fn get_project_by_id(pool: &SqlitePool, id: i64) -> Result<Project, String> {
    let project = sqlx::query_as::<_, Project>(
        "SELECT * FROM projects WHERE id = ?"
    )
    .bind(id)
    .fetch_one(pool)
    .await
    .map_err(|e| format!("获取项目失败: {}", e))?;
    
    Ok(project)
}

pub async fn create_project(pool: &SqlitePool, dto: CreateProjectDto) -> Result<Project, String> {
    let status = dto.status.unwrap_or_else(|| "进行中".to_string());
    
    let project = sqlx::query_as::<_, Project>(
        "INSERT INTO projects (name, description, start_date, end_date, status) VALUES (?, ?, ?, ?, ?) RETURNING *"
    )
    .bind(&dto.name)
    .bind(&dto.description)
    .bind(&dto.start_date)
    .bind(&dto.end_date)
    .bind(&status)
    .fetch_one(pool)
    .await
    .map_err(|e| format!("创建项目失败: {}", e))?;
    
    Ok(project)
}

pub async fn update_project(pool: &SqlitePool, id: i64, dto: UpdateProjectDto) -> Result<Project, String> {
    // 先获取现有项目
    let existing = get_project_by_id(pool, id).await?;
    
    let name = dto.name.unwrap_or(existing.name);
    let description = dto.description.or(existing.description);
    let start_date = dto.start_date.unwrap_or(existing.start_date);
    let end_date = dto.end_date.or(existing.end_date);
    let status = dto.status.unwrap_or(existing.status);
    
    let project = sqlx::query_as::<_, Project>(
        "UPDATE projects SET name = ?, description = ?, start_date = ?, end_date = ?, status = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ? RETURNING *"
    )
    .bind(&name)
    .bind(&description)
    .bind(&start_date)
    .bind(&end_date)
    .bind(&status)
    .bind(id)
    .fetch_one(pool)
    .await
    .map_err(|e| format!("更新项目失败: {}", e))?;
    
    Ok(project)
}

pub async fn delete_project(pool: &SqlitePool, id: i64) -> Result<(), String> {
    sqlx::query("DELETE FROM projects WHERE id = ?")
        .bind(id)
        .execute(pool)
        .await
        .map_err(|e| format!("删除项目失败: {}", e))?;
    
    Ok(())
}
